src/mongo/shell.ts

```diff
@@ -5,120 +5,201 @@
+				this.fireResult(result);
 			} else {
-				buffers.push(b);
+				this._stdoutData += data;
+			}
+		});
+
+		// Monitor STDERR
+		this._mongoShell.stderr.on('data', (buffer: Buffer) => {
+			this._stderrData += buffer.toString();
+		});
+		this._mongoShell.stderr.on('end', () => {
+			if (this._stderrData) {
+				this.fireError(this._stderrData);
 			}
 		});
+	}
 
-		on(this.mongoShell.stderr, 'data', result => this.onResult.fire(result));
-		once(this.mongoShell.stderr, 'close', result => this.onResult.fire(result));
+	private fireResult(result: string): void {
+		this._onResult.fire(<CommandResult>{ result });
 	}
 
-	async useDatabase(database: string): Promise<string> {
-		return this.exec(`use ${database}`);
+	private fireError(error: string | unknown): void {
+		let message: string = typeof error === 'string' ? error : parseError(error).message;
+		if (typeof error === "object" && error["code"] === 'ERR_STREAM_DESTROYED') {
+			message = prematureExitMessage;
+		}
+		this._onResult.fire(<CommandResult>{ errorMessage: message });
+	}
+
+	public async useDatabase(database: string): Promise<string> {
+		return await this.executeScript(`use ${database}`);
 	}
 
-	async exec(script: string): Promise<string> {
+	public async executeScript(script: string): Promise<string> {
+		if (this._mongoShell.killed || this._exited) {
+			throw new Error(prematureExitMessage);
+		}
+
+		this._stdoutData = "";
+		this._stderrData = "";
 		script = this.convertToSingleLine(script);
 		const executionId = this._generateExecutionSequenceId();
+		const shellDetails = `Shell path: "${this._execPath}", arguments: ${JSON.stringify(this._execArgs)}`;
 
 		try {
-			this.mongoShell.stdin.write(script, 'utf8');
-			this.mongoShell.stdin.write(os.EOL);
-			this.mongoShell.stdin.write(executionId, 'utf8');
-			this.mongoShell.stdin.write(os.EOL);
+			this._mongoShell.stdin.write(script, 'utf8');
+			this._mongoShell.stdin.write(os.EOL);
+
+			// Mark end of result by sending the sentinel wrapped in quotes so the console will spit
+			// it back out as a string value
+			this._mongoShell.stdin.write(`"${endOfDataBaseSentinel}${executionId}"`, 'utf8');
+			this._mongoShell.stdin.write(os.EOL);
 		} catch (error) {
-			window.showErrorMessage(error.toString());
+			// Generally if writing to the process' stdin fails it has already exited
+			// with an error, and we will get notification via its stdout. So delay this long
+			// enough for other notifications to be a given a to fire first.
+			setTimeout(() => this.fireError(error), 500);
 		}
 
-		return await new Promise<string>((c, e) => {
-			let executed = false;
-			// timeout setting specified in seconds. Convert to ms for setTimeout
-			const timeout: number = 1000 * vscode.workspace.getConfiguration().get<number>(ext.settingsKeys.mongoShellTimeout);
-			const handler = setTimeout(
+		return await new Promise<string>((resolve, reject) => {
```

Comment:
From personal experience, it's best to put as little code inside `await new Promise` as possible. I've had cases where the code inside of that was throwing an error, which meant `resolve` or `reject` were never called and the feature basically hung with no error.

I think this would be closer to the minimum amount of code:
```typescript
const result = await new Promise<CommandResult>((resolve, reject) => {
	const timeoutHandler = setTimeout(
		() => {
			reject(`Timed out executing MongoDB command "${script}". To use a longer timeout, modify the VS Code 'mongo.shell.timeout' setting.`);
		},
		timeout);

	// Handle result or error from the console (via fireOnResultOrError)
	const disposable = this._onResult.event((r: CommandResult) => {
		clearTimeout(timeoutHandler);

		// Only the first result or error will be processed, all others ignored
		disposable.dispose();
		resolve(r);
	});
});

await delay(100); // Give STDOUT/STDERR a chance to empty

...
```